res/controllers/Traktor-Kontrol-S4-MK2-hid-scripts.js

```diff
@@ -1598,6 +1645,75 @@ TraktorS4MK2.loopSizeSet = function(group) {
 
 TraktorS4MK2.onRecordingChanged = function(value, group, control) {
   TraktorS4MK2.outputChannelCallback(value, "[Recording]", "status");
-  TraktorS4MK2.controller.setOutput("deck1", "!on_air", value*0x7F);
-  TraktorS4MK2.controller.setOutput("deck2", "!on_air", value*0x7F);
+  TraktorS4MK2.controller.setOutput("deck1", "!on_air", value*0x7F, !TraktorS4MK2.controller.freeze_lights);
+  TraktorS4MK2.controller.setOutput("deck2", "!on_air", value*0x7F, !TraktorS4MK2.controller.freeze_lights);
+}
+
+TraktorS4MK2.slipHandler = function(field){
+    // Change on touchdown
+    if (field.value === 0x01){
+      var group = field.id.split(".")[0];
+      if (TraktorS4MK2.controller.play_shift_pressed === 1) {
+          TraktorS4MK2.controller.autoSlipMode[field.group] ^= true;
+          TraktorS4MK2.lightSlipAuto(field.group);
+      } else if (TraktorS4MK2.controller.shift_pressed[group] === 1)   {
+          var prev_range = TraktorS4MK2.controller.tempoRange[field.group];
+          print('Previous range ' + prev_range);
+          if (prev_range <= 0){
+            prev_range = 0;
+            TraktorS4MK2.controller.tempoRange[field.group] = 0;
+          } else if (prev_range > 0) {
+            var new_range = prev_range - 1;
+            var range_value = TraktorS4MK2.controller.tempoRangeList[new_range] / 100;
+            TraktorS4MK2.controller.tempoRange[field.group] = new_range;
+            engine.setValue(field.group, "rateRange", range_value);
+          }
+
+      } else {
+        engine.setValue(field.group, "slip_enabled", !engine.getValue(field.group, "slip_enabled"));
+      }
+    }
+    // Do nothing on touchup
+}
+
+TraktorS4MK2.resetHandler = function(field){
+  var group = field.id.split(".")[0];
+  if (field.value === 0x01){
+    if (TraktorS4MK2.controller.shift_pressed[group] === 1){
+      var prev_range = TraktorS4MK2.controller.tempoRange[field.group];
+      print('Previous range ' + prev_range);
+      if (prev_range >= TraktorS4MK2.controller.tempoRangeList.length - 1){
+        prev_range = TraktorS4MK2.controller.tempoRangeList.length - 1;
+        TraktorS4MK2.controller.tempoRange[field.group] = prev_range;
+      } else {
+        var new_range = prev_range + 1;
+        var range_value = TraktorS4MK2.controller.tempoRangeList[new_range] / 100;
+        TraktorS4MK2.controller.tempoRange[field.group] = new_range;
+        engine.setValue(field.group, "rateRange", range_value);
```

Comment:
The variable names in this function are confusing like the `TraktorS4Mk2.slipHandler` function above.